src/ir/gc-type-utils.h

```diff
@@ -104,7 +104,19 @@ inline EvaluationResult evaluateCastCheck(Type refType, Type castType) {
   }
 
   auto castHeapType = castType.getHeapType();
-  auto refIsHeapSubType = HeapType::isSubType(refHeapType, castHeapType);
+
+  // Check whether a value of type `a` is known to also have type `b`, assuming
+  // it is non-null.
+  auto isHeapSubtype = [](Type a, Type b) {
+    // TODO: Use information from a subtypes analysis, if available.
+    if (!a.getHeapType().isBasic() && !a.getHeapType().isOpen()) {
+      a = a.with(Exact);
+    }
+    // Ignore nullability.
+    return Type::isSubType(a.with(NonNullable), b.with(NonNullable));
+  };
```

Comment:
I'm still not following this comment+code. If `a` is a user-defined heap type, why does it matter if it is open or not?

E.g. I would expect `isHeapSubtype(child, parent)` to return true regardless of the openness of `child`, since `child` "also has type" `parent`. Or do I not understand what "also has type" means?

Reply:
The "heap type" being checked here includes exactness, which is why we are actually comparing types and selectively ignoring nullability.

We're not just comparing the static type of the cast input with the cast target type. Instead, we're comparing the most specific safe approximation of the _dynamic_ type of the cast input we can get with the cast target. If we know the value has dynamic heap type $foo, and $foo is final and cannot have subtypes, then we also know that the value must have dynamic heap type (exact $foo) because it cannot have any other type and still be a $foo.

Reply:
Sorry, what is a "dynamic type"?

Reply:
The type of the value at runtime.

Reply:
Thanks, now I can follow the code here.

I wrote a comment suggestion above.